### PR TITLE
Fix txg_wait_open() load average inflation

### PR DESCRIPTION
**Pull request by behlendorf**
### Motivation and Context

Issue #8550.  Fix inflated load average.

### Description

When waiting in `txg_wait_open()` sleep interruptibly to prevent inflating the load average.  The code is already structure such that it's safe to be woken early by a signal.

### How Has This Been Tested?

Prior to this change when `autotrim=on` a pool with a single top-level vdev would have a load average of approximately __1__ even when idle.  After this change the load average is __0.01__.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/txg.c [hunk 1]
```diff
@@ -715,7 +715,7 @@ txg_wait_open(dsl_pool_t *dp, uint64_t txg, boolean_t should_quiesce)
 	    txg, tx->tx_quiesce_txg_waiting, tx->tx_sync_txg_waiting);
 	while (tx->tx_open_txg < txg) {
 		cv_broadcast(&tx->tx_quiesce_more_cv);
-		cv_wait(&tx->tx_quiesce_done_cv, &tx->tx_sync_lock);
```
Comment (ahrens) [inline module/zfs/txg.c]: Could you add a comment explaining why we are doing cv_wait_sig() but not handling the signal?  Would we want to use _sig in other places as well (e.g. txg_wait_synced())?

Comment (behlendorf) [reply module/zfs/txg.c]: Sure, I'll add a comment.  We've needed to make this same change, for the same reason, in a few other places as well so I'll audit those and add comments as needed.

Regarding, `txg_wait_synced()` I think we should consider switching it to `cv_wait_io()`.  This was added for Linux and is identical to `cv_wait()`, except it calls `io_schedule()` internally to indicate the blocked process is waiting on IO.  If we do decide that's the right behavior, I'd suggest we do it in a follow up PR.

Comment (behlendorf) [reply module/zfs/txg.c]: Refreshed.  I annotated some of callers we converted from`cv_wait()` to `cv_wait_sig()` but it felt gratuitous to add the same boiler plate comment to them all.   I think we should consider adding a `porting.txt` document which explains some of these subtle differences between platforms.

**Comment by pfmooney** [comment]
I'll repeat my apprehension from [here](https://github.com/behlendorf/zfs/commit/024e5a5adf9ff8b983276114c8d4c450e5b1d164#r33007075): Converting a `cv_wait` to a `cv_wait_sig` without checking or handling the return value is a recipe for a tight spin on illumos-based platforms, should a signal ever be queued on that thread.

**Comment by behlendorf** [comment]
@pfmooney you're right this code as it stands is a recipe for a tight spin on Illumos, and it is not portable. However, it is currently safe for ZoL in this context because we ignore all signals in the txg_quiesce thread.  It should only affect the load average.

```
# cat /proc/8639/comm 
txg_quiesce

# cat /proc/8639/status | grep Sig
SigQ:	3/15651
SigPnd:	0000000000000000
SigBlk:	0000000000000000
SigIgn:	ffffffffffffffff
SigCgt:	0000000000000000
```

The Linux kernel doesn't provide any primitives for condition variables (alternate interfaces are preferred).  This means when we implemented the cv interfaces we only implemented what was absolutely needed.  At the time that didn't include `cv_wait_sig()` so we only have a partial implementation.

I completely agree that we need to finish implementing this functionality so we can use it in a portable way.  In fact, we're going to need to in order to port illumos 9425.  So that would be a good time to tackle this.

**Comment by ahrens** [comment]
If we are going to change txg_wait_open() to use cv_wait_sig, shouldn't we also change txg_wait_synced()?

> it is currently safe for ZoL in this context because we ignore all signals

txg_wait_open() and txg_wait_synced() are called from syscalls (e.g. via dmu_free_long_range_impl()), in which case I don't think we are ignoring signals.

**Comment by behlendorf** [comment]
> If we are going to change txg_wait_open() to use cv_wait_sig, shouldn't we also change txg_wait_synced()?

Good question.  I think it depends on how we want this wait time to be reported.  When calculating the system load Linux counts processes in an uninterruptible sleep state as __system__ time (along with the running state).  My understanding is other Unix-like systems report things differently.

For `txg_wait_synced()` I think this mostly makes sense since those process are blocked waiting on work to get done.  Alternately we could report this as __iowait__ time, instead of __system__ time, using the `cv_wait_io()` wrapper which was added for this purpose.

My feeling is __iowait__ is probably more correct, and useful when profiling a system.  It would also be a user visible change to long standing behavior.  I don't think we'd want to use `cv_wait_sig()` here.

My primary concern is that there is an expectation that an idle system should have a load average of effectively zero.  And with the autotrim enabled and waiting in `txg_wait_open()` that wasn't the case.

> txg_wait_open() and txg_wait_synced() are called from syscalls (e.g. via dmu_free_long_range_impl()), in which case I don't think we are ignoring signals.

Good point.  Why don't I update this to only use `cv_wait_sig()` when `should_quiesce == B_FALSE`.  The autotrim threads are the only callers which call `cv_wait_sig()` this way, and I can update the comment.  Then we can follow up with a change to fully implement `cv_wait_sig()`.

**Comment by richardelling** [comment]
accounting against iowait is better

File: module/zfs/zfs_sysfs.c [hunk 2]
```diff
@@ -356,38 +356,68 @@ pool_property_show(struct kobject *kobj, struct attribute *attr, char *buf)
  * async calls (i.e. a sync flag). If this mechanism were in place at that
  * time, we could have added a 'channel_program_async' to this list.
  */
-static const char *zfs_features[]  = {
-	/* --> Add new kernel features here (post ZoL 0.8.0) */
-	"initialize",
-	"trim",
+static const char *zfs_kernel_features[]  = {
+	/* --> Add new kernel features here */
+	"com.delphix.vdev_initialize",
```
Comment (ahrens) [inline module/zfs/zfs_sysfs.c]: Looks like you accidentally included Don's commit `features.kernel layout should match features.pool`

**Comment by ahrens** [comment]
Thanks for the link to Brendan's blog post, that was informative.

It makes sense for the txg_sync_thread() and txg_quiesce_thread() to not contribute to the load average when they are truly idle and just waiting for more work (i.e. are in txg_thread_wait()).  Same goes for other internal threads that are waiting for work, e.g. taskq threads.  I think it would be more clear and less fragile to have a function specifically for this purpose, rather than using this side-effect of cv_wait_sig() (e.g. introduce a new cv_wait_idle()).  Can't say that you have to do this now just because you're adding one more cv_wait_sig() though.

I agree that txg_wait_synced() and txg_wait_open(quiesce=TRUE) should count as waiting waiting for i/o (since they are), and thus use cv_wait_io().  This change is related but orthogonal to the issue you're fixing here.  I think that we should strongly consider doing this despite the fact that it would change user-visible behavior.

So in conclusion, I'm fine with the current proposal, but it would be nice to also make the 2 additional changes mentioned above.

File: module/zfs/zfs_sysfs.c (hunk 2 above)
Comment (behlendorf) [reply module/zfs/zfs_sysfs.c]: Whoops.   Fixed.

**Comment by behlendorf** [comment]
Refreshed.  `txg_wait_synced()` and `txg_wait_open()` were updated accordingly.  I'm happy to follow up with a PR which adds a new `cv_wait_idle()` function.

**Comment by lundman** [comment]
Wait, what does `cv_wait_idle()` do? I would have to implement it as well :)

**Comment by behlendorf** [comment]
The proposal is to create a `cv_wait_idle()` which behaves exactly like `cv_wait()` _except_ it's accounted as idle instead of system time (we're effectively exploiting cv_wait_sig() for this behavior today).  I believe on pretty much everything Unix-like other than Linux you could just define `cv_wait_idle()` as `cv_wait()`.  But I'd be curious to know the situation on other platforms!
